### ISSUE TITLE
File-Scoped Types With The Same Name Trigger CA1708 

### ISSUE THREAD

**Issue by JasonBock** [dotnet/roslyn-analyzers#6514]
### Analyzer

**Diagnostic ID**: [CA1708](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1708): `Identifiers should differ by more than case`

### Analyzer source

**SDK**: .NET 7

**Version**: 7.0.3

### Describe the bug

Basically what I say in the title. File-scoped types with the same name are triggering this issue.

### Steps To Reproduce

Create a console application, targeting .NET 7, and ensure the .csproj looks like this:

```xml
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup>
		<AnalysisMode>all</AnalysisMode>
		<TreatWarningsAsErrors>true</TreatWarningsAsErrors>
		<OutputType>Exe</OutputType>
		<TargetFramework>net7.0</TargetFramework>
		<ImplicitUsings>enable</ImplicitUsings>
		<Nullable>enable</Nullable>
	</PropertyGroup>
</Project>
```

Add two .cs files to the project, call them whatever you want (Class1.cs, Class2.cs), and put the following code into the files:

Class1.cs
```csharp
namespace ConsoleApp1;

file static class GeneratedName
{
    public const string Name = "Method";
}

public static class MethodName
{
    public static string RetrieveName() => GeneratedName.Name;
}
```

Class2.cs
```csharp
namespace ConsoleApp1;

file static class GeneratedName
{
    public const string Name = "Type";
}

public static class TypeName
{
    public static string RetrieveName() => GeneratedName.Name;
}
```

### Expected behavior

No issues should be created.

### Actual behavior

CA1708 is created on a build, something like this:

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	CA1708	Names of 'Types' and 'ConsoleApp1.GeneratedName, ConsoleApp1.GeneratedName' should differ by more than case	ConsoleApp1	C:\Users\jason\source\repos\ConsoleApp1\ConsoleApp1\ConsoleApp1.csproj	1	Active
```

**Comment by mpidash** [comment]
The version of `Microsoft.CodeAnalysis` that `Microsoft.CodeAnalysis.NetAnalyzers` uses (`3.11.0`) does not yet expose the [`INamedTypeSymbol.IsFileLocal`](https://github.com/dotnet/roslyn/blob/f5b6c715a742c56b7cc672e47385508fb4df98cc/src/Compilers/Core/Portable/Symbols/INamedTypeSymbol.cs#L59-L62) property:

```C#
public interface INamedTypeSymbol : ITypeSymbol
{
...

  /// <summary>
  /// Indicates the type is declared in source and is only visible in the file it is declared in.
  /// </summary>
  bool IsFileLocal { get; }

...
}
```

Is there a workaround to determine, if a symbol is file-scoped?

**Comment by svick** [comment]
When fixing this, I think the error message should be fixed as well. It should probably be:

> Names of types 'ConsoleApp1.GeneratedName, ConsoleApp1.GeneratedName' should differ by more than case

**Comment by RenderMichael** [comment]
Could I get assigned this? I think I have a working solution

> Is there a workaround to determine, if a symbol is file-scoped?

`item.MetadataName.Contains('<')` should work fine as a workaround.

**Comment by mavasani** [comment]
> Could I get assigned this? I think I have a working solution
> 
> > Is there a workaround to determine, if a symbol is file-scoped?
> 
> `item.MetadataName.Contains('<')` should work fine as a workaround.

@333fred is this a reasonable workaround?

@RenderMichael another option is to define this property in the lightup layer and use reflection.

I am fine with either approach here.

**Comment by 333fred** [comment]
I would prefer the lightup approach.

**Comment by RenderMichael** [comment]
Using the `item.MetadataName.Contains('<')` approach doesn't just hit file-scoped types, but any compiler-generated types like anonymous types. It just so happens that file-scoped types are exposed and named in a way other compiler-generated types don't.

In fact, this analyzer could compare type names using `MetadataName` to begin with, then this problem wouldn't exist.

**Comment by RenderMichael** [comment]
Opened PR #7190 using the light-up approach.
